File: Chapter 02/BridgeDesignPatternExample.java
abstract class Corgi {
    protected Color color;

    public Corgi(Color color) {
        this.color = color;
    }

    abstract void paint();
}


class Tail extends Corgi {
    private int length;

    public Tail(Color color, int length) {
        super(color);
        this.length = length;
    }

    void paint() {
        System.out.print("Painting a tail with ");
        color.applyColor();
        System.out.println("Length: " + length);
    }
}


interface Color {
    void applyColor();
}

class Black implements Color {
    public void applyColor() {
        System.out.print("Black Color");
    }
}

class Gold implements Color {
    public void applyColor() {
        System.out.print("Gold Color");
    }
}


public class BridgeDesignPatternExample {
    public static void main(String[] args) {
        Corgi blackTail = new Tail(new Black(), 15);
        Corgi goldTail = new Tail(new Gold(), 18);

        blackCorgi.paint();
        goldCorgi.paint();
    }
}

File: Chapter 02/BridgeDesignPatternExample.py
class Corgi:
    def __init__(self, color):
        self.color = color
    def paint(self):
        pass
class Tail(Corgi):
    def __init__(self, color, length):
        super().__init__(color)
        self.length = length
    def paint(self):
        print(f"Painting a tail with {self.color.apply_color()} - Length: {self.length}")
class Color:
    def apply_color(self):
        pass
class Black(Color):
    def apply_color(self):
        return "Black Color"
class Gold(Color):
    def apply_color(self):
        return "Gold Color"
if __name__ == "__main__":
    black_tail = Tail(Black(), 15)
    gold_tail = Tail(Gold(), 18)

    black_tail.paint()
    gold_tail.paint()
